File: packages/python/scripts/perf_smoke.py
from __future__ import annotations

import os
import shutil
import tempfile
import time

from harness_detect import CheckOptions, detect_harnesses, list_harnesses


def main() -> None:
    iterations = read_iterations()
    harness_count = len(list_harnesses())
    root_dir = tempfile.mkdtemp(prefix="harness-detect-perf-")
    try:
        options = build_options(root_dir)
        warmup = detect_harnesses(options)
        if len(warmup) != harness_count:
            raise RuntimeError(
                f"warmup detect count mismatch: got {len(warmup)}, want {harness_count}"
            )

        started_at = time.perf_counter()
        for index in range(iterations):
            results = detect_harnesses(options)
            if len(results) != harness_count:
                raise RuntimeError(
                    f"detect count mismatch on iteration {index + 1}: got {len(results)}, want {harness_count}"
                )
        elapsed = time.perf_counter() - started_at
        elapsed_ms = elapsed * 1000
        ops_per_sec = iterations / elapsed
        print(
            f"python perf: {iterations} iterations, {ops_per_sec:.2f} ops/sec, "
            f"{elapsed_ms:.2f} ms elapsed, {harness_count} harnesses/run, "
            "PATH=empty, hermetic env"
        )
    finally:
        shutil.rmtree(root_dir, ignore_errors=True)


def read_iterations() -> int:
    raw = os.environ.get("HARNESS_DETECT_PERF_ITERATIONS")
    if raw is None:
        return 250
    try:
        parsed = int(raw)
    except ValueError as error:
        raise ValueError(
            f"HARNESS_DETECT_PERF_ITERATIONS must be a positive integer, got {raw}"
        ) from error
    if parsed <= 0:
        raise ValueError(
            f"HARNESS_DETECT_PERF_ITERATIONS must be a positive integer, got {raw}"
        )
    return parsed


def build_options(root_dir: str) -> CheckOptions:
    home = os.path.join(root_dir, "home")
    cwd = os.path.join(root_dir, "cwd")
    for directory in (
        home,
        cwd,
        os.path.join(home, ".config"),
        os.path.join(home, ".local", "share"),
        os.path.join(home, ".local", "state"),
        os.path.join(home, ".cache"),
    ):
        os.makedirs(directory, exist_ok=True)

    env = {
        "HOME": home,
        "PATH": "",
        "XDG_CONFIG_HOME": os.path.join(home, ".config"),
        "XDG_DATA_HOME": os.path.join(home, ".local", "share"),
        "XDG_STATE_HOME": os.path.join(home, ".local", "state"),
        "XDG_CACHE_HOME": os.path.join(home, ".cache"),
        "CODEX_HOME": os.path.join(root_dir, "overrides", "codex"),
        "CLAUDE_CONFIG_DIR": os.path.join(root_dir, "overrides", "claude"),
        "GEMINI_CLI_HOME": os.path.join(root_dir, "overrides", "gemini-home"),
        "OPENCODE_CONFIG_DIR": os.path.join(root_dir, "overrides", "opencode-config"),
        "GOOSE_PATH_ROOT": os.path.join(root_dir, "overrides", "goose"),
        "CLINE_DIR": os.path.join(root_dir, "overrides", "cline"),
        "Q_CLI_DATA_DIR": os.path.join(root_dir, "overrides", "amazon-q"),
        "COPILOT_HOME": os.path.join(root_dir, "overrides", "copilot"),
        "COPILOT_CACHE_HOME": os.path.join(root_dir, "overrides", "copilot-cache"),
        "AMP_DATA_HOME": os.path.join(root_dir, "overrides", "amp-data"),
        "HERMES_HOME": os.path.join(root_dir, "overrides", "hermes"),
        "OPENCLAW_HOME": os.path.join(root_dir, "overrides", "openclaw"),
        "OPENCLAW_STATE_DIR": os.path.join(root_dir, "overrides", "openclaw-state"),
        "AUTOGENSTUDIO_APPDIR": os.path.join(root_dir, "overrides", "autogenstudio"),
    }
    return CheckOptions(env=env, cwd=cwd)


if __name__ == "__main__":
    main()

File: packages/typescript/scripts/perf-smoke.js
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import process from "node:process";
import { performance } from "node:perf_hooks";

import { detectHarnesses, listHarnesses } from "../src/index.js";

function makeDir(dirPath) {
  fs.mkdirSync(dirPath, { recursive: true });
}

function buildEnv(rootDir) {
  const home = path.join(rootDir, "home");
  const cwd = path.join(rootDir, "cwd");

  makeDir(home);
  makeDir(cwd);
  makeDir(path.join(home, ".config"));
  makeDir(path.join(home, ".local", "share"));
  makeDir(path.join(home, ".local", "state"));
  makeDir(path.join(home, ".cache"));

  return {
    cwd,
    env: {
      HOME: home,
      PATH: "",
      XDG_CONFIG_HOME: path.join(home, ".config"),
      XDG_DATA_HOME: path.join(home, ".local", "share"),
      XDG_STATE_HOME: path.join(home, ".local", "state"),
      XDG_CACHE_HOME: path.join(home, ".cache"),
      CODEX_HOME: path.join(rootDir, "overrides", "codex"),
      CLAUDE_CONFIG_DIR: path.join(rootDir, "overrides", "claude"),
      GEMINI_CLI_HOME: path.join(rootDir, "overrides", "gemini-home"),
      OPENCODE_CONFIG_DIR: path.join(rootDir, "overrides", "opencode-config"),
      GOOSE_PATH_ROOT: path.join(rootDir, "overrides", "goose"),
      CLINE_DIR: path.join(rootDir, "overrides", "cline"),
      Q_CLI_DATA_DIR: path.join(rootDir, "overrides", "amazon-q"),
      COPILOT_HOME: path.join(rootDir, "overrides", "copilot"),
      COPILOT_CACHE_HOME: path.join(rootDir, "overrides", "copilot-cache"),
      AMP_DATA_HOME: path.join(rootDir, "overrides", "amp-data"),
      HERMES_HOME: path.join(rootDir, "overrides", "hermes"),
      OPENCLAW_HOME: path.join(rootDir, "overrides", "openclaw"),
      OPENCLAW_STATE_DIR: path.join(rootDir, "overrides", "openclaw-state"),
      AUTOGENSTUDIO_APPDIR: path.join(rootDir, "overrides", "autogenstudio")
    }
  };
}

function readIterations() {
  const raw = process.env.HARNESS_DETECT_PERF_ITERATIONS;
  if (!raw) {
    return 250;
  }

  const parsed = Number.parseInt(raw, 10);
  if (!Number.isFinite(parsed) || parsed <= 0) {
    throw new Error(`HARNESS_DETECT_PERF_ITERATIONS must be a positive integer, got ${raw}`);
  }

  return parsed;
}

const iterations = readIterations();
const harnessCount = listHarnesses().length;
const rootDir = fs.mkdtempSync(path.join(os.tmpdir(), "harness-detect-perf-"));

try {
  const { cwd, env } = buildEnv(rootDir);
  const warmup = detectHarnesses({ env, cwd });
  if (warmup.length !== harnessCount) {
    throw new Error(`warmup detect count mismatch: got ${warmup.length}, want ${harnessCount}`);
  }

  const startedAt = performance.now();
  for (let i = 0; i < iterations; i += 1) {
    const results = detectHarnesses({ env, cwd });
    if (results.length !== harnessCount) {
      throw new Error(`detect count mismatch on iteration ${i + 1}: got ${results.length}, want ${harnessCount}`);
    }
  }
  const elapsedMs = performance.now() - startedAt;
  const opsPerSec = iterations / (elapsedMs / 1000);

  console.log(`typescript perf: ${iterations} iterations, ${opsPerSec.toFixed(2)} ops/sec, ${elapsedMs.toFixed(2)} ms elapsed, ${harnessCount} harnesses/run, PATH=empty, hermetic env`);
} finally {
  fs.rmSync(rootDir, { recursive: true, force: true });
}
